### PR TITLE
nmcli: suggest new routes4 and routes6 format

### LINKED ISSUE #4131: nmcli reports changed for routes4 parameter that includes attributes
### Summary

When using nmcli with the routes4 parameter that includes attributes for the routes it always reports as changed.   A different format needs to be used for the check then when setting the parameter.

In particular when using the "table" attribute, when checking the value the following format needs to be used with "ip =" and "nh =" for the check to not show change, but when configuring the interface those cannot be included.

For example if "0.0.0.0/0 192.168.138.1 table=201" is used then it will not be idempotent and show changed.
--- before
+++ after
@@ -10 +10 @@
-- '{ ip = 0.0.0.0/0, nh = 192.168.138.1 table=201 }'
+- 0.0.0.0/0 192.168.138.1 table=201

If "{ ip = 0.0.0.0/0, nh = 192.168.138.1 table=201 }" used then it will correctly show as not changed if set, but if not already set it needs to be in the format "0.0.0.0/0 192.168.138.1 table=201" when setting.

### Issue Type

Bug Report

### Component Name

community.general.nmcli

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.11.5]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/etc/ansible/collections/ansible_collections/tfdm/utils/plugins/modules', '/usr/share/ansible/collections/ansible_collections/tfdm/utils/plugins/modules', '/home/tfdmseed/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.8/site-packages/ansible
  ansible collection location = /home/tfdmseed/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.8.6 (default, Jan 22 2021, 11:41:28) [GCC 8.4.1 20200928 (Red Hat 8.4.1-1)]
  jinja version = 3.0.2
  libyaml = True
```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general
# /usr/share/ansible/collections/ansible_collections
Collection        Version
----------------- -------
community.general 3.7.0

```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

RHEL 8

### Steps to Reproduce

See routes4 parameter.

```yaml (paste below)
# Run in check_mode first with routes4 paramater formatted for check mode
# This is needed due to bug in nmcli where it checks routes4 in different format
- name: Check nmcli configure interfaces
  community.general.nmcli:
    conn_name: "{{ interface }}"
    ifname: "{{ interface }}"
    vlandev: "{{ vlandev }}"
    vlanid: "{{ vlanid }}"
    type: "{{ type }}"
    ip4: "{{ ip4 }}"
    gw4: "{{ gw4 }}"
    method4: "{{ ipv4_method }}"
    never_default4: "{{ never_default4 }}"
    method6: "{{ ipv6_method }}"
    routes4: "{ ip = 0.0.0.0/0, nh = 192.168.138.1 table=201 }"
    routing_rules4: "{{ routing_rules4 }}"
    state: present
    autoconnect: yes
  check_mode: yes
  register: nmcli_result_check

- name: Configure interfaces
  community.general.nmcli:
    conn_name: "{{ interface }}"
    ifname: "{{ interface }}"
    vlandev: "{{ vlandev }}"
    vlanid: "{{ vlanid }}"
    type: "{{ type }}"
    ip4: "{{ ip4 }}"
    gw4: "{{ gw4 }}"
    method4: "{{ ipv4_method }}"
    never_default4: "{{ never_default4 }}"
    method6: "{{ ipv6_method }}"
    # Use routes4 format depending on if changes are required, to handle idempotency
    routes4: "{{ '0.0.0.0/0 192.168.138.1 table=201' if nmcli_result_check.changed else '{ ip = 0.0.0.0/0, nh = 192.168.138.1 table=201 }'}}"
    routing_rules4: "{{ routing_rules4 }}"
    state: present
    autoconnect: yes
  register: nmcli_result

```

### Expected Results

The "0.0.0.0/0 192.168.138.1 table=201" format should not report a change if already set.   It should be able to be used for both the check and configuring route4 parameter.

### Actual Results

```console (paste below)
{ ip = 0.0.0.0/0, nh = 192.168.138.1 table=201 }' is required when checking interface in order for it to not report change.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by haddystuff**
##### SUMMARY
This PR fix #4131 by adding new routes4 and routes6 format.
My idea is to make adding new routes looks like this:
```
    - name: Create dummy connection
      community.general.nmcli:
        conn_name: dummy_con0
        ifname: dummy0
        ip4: 10.0.0.1/24
        state: present
        type: dummy
        mtu: 1500
        routes4: 
          - ip: 10.1.0.0/24
            next_hop: 10.0.0.2
            metric: 100
            cwnd: 240000
            mtu: 1500
            onlink: yes
            tos: 7
        routes6:
          - ip: 2001:db8:c::9f:35/128
            next_hop: 2001:db8:c::9f:36
            metric: 100
            table: 201
            cwnd: 240000
            mtu: 1500
            onlink: yes
```

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/modules/net_tools/nmcli.py

##### ADDITIONAL INFORMATION
Any suggestions related to new parameters or methods naming highly appreciated.

### CODE REVIEW

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:95]: This change is problematic, because it is a breaking change. Current uses of these parameters will fail once this is merged. That's why we cannot merge this (as-is).

What is possible:
1. Make this `elements: raw` and write code to make sure that these are either strings or dictionaries in the exact format you require.
2. Alternatively, add new options that are lists of dicts and make these mutually exclusive with the old options.

Both alternatives aren't great. 2. allows to properly document the format for the new options, while 1. does not allow to do this (`suboptions` only work for `type=dict` or `type=list`/`elements=dict`).

**Review by felixfontein — CHANGES_REQUESTED** [review]
Thanks for your contribution! Unfortunately this is a breaking change.

File: plugins/modules/net_tools/nmcli.py
Comment (haddystuff) [reply plugins/modules/net_tools/nmcli.py:95]: I really like the second idea, but should we make new option mutually exclusive? Maybe i can make them work simultaneously.

Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:95]: Not making it mutually exclusive usually makes the code more complicated, because then you somehow have to join the two lists, care about ordering, etc. If it's just one or the other, there's no ambiguity what can happen.

**Comment by haddystuff** [comment]
I added new 2 new options and made them mutually exclusive with old ones

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:100]:
```suggestion
            - The list of IPv4 routes.
            - Mutually exclusive with I(routes4).
```
Also mention that this one exists in the `routes4` documentation.

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:106]:
```suggestion
                    - IP or prefix of route.
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:107]:
```suggestion
                    - Use the format C(192.0.3.0/24).
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:112]: There should be some description as well.
```suggestion
                    - Use the format C(192.0.2.1).
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:121]:
```suggestion
                    - The default depends on C(ipv4.route-table).
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:137]:
```suggestion
                    - The Type Of Service.
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:1709]:
```suggestion
        if route.get('next_hop') is not None:
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:1711]:
```suggestion
        if route.get('metric') is not None:
```

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:235]: Shouldn't this probably be
```suggestion
                    - The default depends on C(ipv6.route-table).
```
?

File: plugins/modules/net_tools/nmcli.py
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:94]: How about:
```suggestion
            - Use the format C(192.0.3.0/24 192.0.2.1).
            - To specify more complex routes, use the I(routes4_extended) option.
```
(and something similar for `routes6`?)

**Review by felixfontein — APPROVED** [review]
Looks good to me! (Though I'm not an nmcli user...)

**Comment by felixfontein** [comment]
If nobody complains in a couple of days, I'll merge.

**Comment by felixfontein** [comment]
@haddystuff thanks for implementing this!

**Pull request merged**
